sway/desktop/render.c

```diff
@@ -72,6 +72,82 @@ enum corner_location get_rotated_corner(enum corner_location corner_location,
 	return corner_location;
 }
 
+static struct wlr_box get_monitor_box(struct wlr_output *output) {
+	int width, height;
+	wlr_output_transformed_resolution(output, &width, &height);
+	struct wlr_box monitor_box = { 0, 0, width, height };
+	return monitor_box;
+}
+
+// Adjust the box position when switching the workspace
+static void adjust_box_to_workspace_offset(struct wlr_box *box,
+		bool on_focused_workspace, struct sway_output *output) {
+	float scroll_percent = output->workspace_scroll.percent;
+
+	int ws_dimen;
+	int *box_coord;
+	switch (output->workspace_scroll.direction) {
+	case SWIPE_GESTURE_DIRECTION_NONE:
+		return;
+	case SWIPE_GESTURE_DIRECTION_HORIZONTAL:
+		ws_dimen = output->width;
+		box_coord = &box->x;
+		break;
+	case SWIPE_GESTURE_DIRECTION_VERTICAL:
+		ws_dimen = output->height;
+		box_coord = &box->y;
+		break;
+	}
+
+	*box_coord -= ws_dimen * scroll_percent;
+	if (!on_focused_workspace) {
+		if (scroll_percent > 0) {
+			*box_coord += ws_dimen;
+		} else if (scroll_percent < 0) {
+			*box_coord -= ws_dimen;
+		}
+	}
+}
+
+// Clips the rendered damage to the workspace region.
+// Fixes containers being rendered across workspaces while switching.
+static void adjust_damage_to_workspace_bounds(pixman_region32_t *damage,
+		bool on_focused_workspace, struct sway_output *output) {
+	float scale = output->wlr_output->scale;
+	float scroll_percent = output->workspace_scroll.percent;
+	int x = 0, y = 0;
+
+	int ws_dimen;
+	int *coord;
+	switch (output->workspace_scroll.direction) {
+	case SWIPE_GESTURE_DIRECTION_NONE:
+		return;
+	case SWIPE_GESTURE_DIRECTION_HORIZONTAL:
+		ws_dimen = output->width;
+		coord = &x;
+		break;
+	case SWIPE_GESTURE_DIRECTION_VERTICAL:
+		ws_dimen = output->height;
+		coord = &y;
+		break;
+	}
```

Comment:
this bit is confusing to me, may need some clarification - what do x and y represent if they remain at 0 - and why does `coord` depend on them?

Reply:
Oh yeah, that is a bit confusing… xD

It's mostly there to remove duplicate code due to the gesture either needing to adjust the X or Y coordinates depending on the gesture direction. Do you have any other way of solving this without duplicating the code?

Maybe just add the x and y coords as args?